flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,449 @@
+
+CapCons are a non-storable object.
+
+Capabilities also have an address field and a target field. The target field is the storage path of the targeted object. Creating a capability from a capability should be illegal.
+
+The definition of the `CapabilityController` .
+
+```cadence
+// CapabilityController can be retrieved via:
+// - AuthAccount::getControllers(path: StoragePath): [CapabilityController]
+// - AuthAccount::getController(capabilityID: UInt64): CapabilityController?
+struct CapabilityController {
+   // The block height when the capability was created.
+   let issueHeight: UInt64
+   // The Type of the capability.
+   let CapabilityType: Type
+   // The id of the related capability.
+   // This is the UUID of the created capability.
+   let capabilityID: UInt64
+
+   // Is the capability revoked.
+   fun isRevoked(): Bool
+   // The target storage path.
+   fun target(): StoragePath
+   // Revoke the capability.
+   fun revoke()
+   // Retarget the capability.
+   // This would move the CapCon from one CapCon array to another.
+   // If the new target is not a valid target this will panic.
+   fun retarget(target: StoragePath)
+}
+```
+
+The auth account would get new methods to get CapCons in order to manage capabilities:
+
+```cadence
+// Get all capability controllers for capabilities that target this storage path
+fun getControllers(path: StoragePath): [CapabilityController]
+// Get capability controller for capability with the specified id
+// If the id does not reference an existing capability
+// or the capability does not target a storage path on this address, return nil
+fun getController(capabilityID: UInt64): CapabilityController?
+```
+
+Some methods would be removed from the `AuthAccount` object as they are no longer needed:
+
+```cadence
+fun link<T: &Any>(_ newCapabilityPath: CapabilityPath, target: Path): Capability<T>?
+fun getLinkTarget(_ path: CapabilityPath): Path?
+fun unlink(_ path: CapabilityPath)
+```
+
+The method `getCapability` would be renamed to `issueCapability `to reflect the fact that a new capability is created every time.
+
+And also from the PublicAccount:
+
+```cadence
+fun getCapability<T>(_ path: PublicPath): Capability<T>
+fun getLinkTarget(_ path: CapabilityPath): Path?
+```
+
+This would remove all references to `Link`s.
+
+### Borrow Capability
+
+. The `borrowCapability `method would be added to the PublicAccount which would be a shorthand to how we currently get the capability then call `borrow` on it.
+
+```cadence
+// this:
+
+var publicAccount = getAccount(issuerAddress)
+var cap = getCapability<&{CounterContract.HasCount}>(/public/counter)
+var countCap = cap.borrow()!
+countCap.count
+
+
+// become this:
+
+let publicAccount = getAccount(issuerAddress)
+let countCap = publicAccount.borrowCapability<&{HasCount}>(/public/hasCount)!
```

Comment:
As I understood the methods, I thought this would be either

```
let countCap: Capability<&{HasCount}> = publicAccount.issueCapability<&{HasCount}>(/public/hasCount)!
```

or

```
let countRef: &AnyResource{HasCount} = publicAccount.borrowCapability<&{HasCount}>(/public/hasCount)!
```

Reply:
@sisyphusSmiling Agreed, that is probably a typo:

```suggestion
let countRef = publicAccount.borrowCapability<&{HasCount}>(/public/hasCount)!
```